src/instructions/segmentation.rs

```diff
@@ -1,210 +1,278 @@
+                {
+                    let val: u64;
+                    asm!(concat!("rd", $name, "base {}"), out(reg) val, options(nomem, nostack, preserves_flags));
+                    VirtAddr::new_unsafe(val)
+                }
+                #[cfg(not(feature = "inline_asm"))]
+                VirtAddr::new_unsafe(crate::asm::$asm_rd())
+            }
+
+            unsafe fn write_base(base: VirtAddr) {
+                #[cfg(feature = "inline_asm")]
+                asm!(concat!("wr", $name, "base {}"), in(reg) base.as_u64(), options(nostack, preserves_flags));
 
-    #[cfg(not(feature = "inline_asm"))]
-    crate::asm::x86_64_asm_load_es(sel.0);
+                #[cfg(not(feature = "inline_asm"))]
+                crate::asm::$asm_wr(base.as_u64());
+            }
+        }
+    };
 }
 
-/// Reload fs segment register.
+/// Code Segment
 ///
-/// ## Safety
-///
-/// This function is unsafe because the caller must ensure that `sel`
-/// is a valid fs segment descriptor.
-#[inline]
-pub unsafe fn load_fs(sel: SegmentSelector) {
-    #[cfg(feature = "inline_asm")]
-    asm!("mov fs, {0:x}", in(reg) sel.0, options(nostack, preserves_flags));
+/// The segment base and limit are unused in 64-bit mode. Only the L (long), D
+/// (default operation size), and DPL (descriptor privilege-level) fields of the
+/// descriptor are recognized. So chaning the segment register can be used to
+/// change privilege level or enable/disable long mode.
+#[derive(Debug)]
+pub struct CS;
+impl Segment for CS {
+    get_reg_impl!("cs", x86_64_asm_get_cs);
 
-    #[cfg(not(feature = "inline_asm"))]
-    crate::asm::x86_64_asm_load_fs(sel.0);
+    unsafe fn set_reg(sel: SegmentSelector) {
+        // Note this is special since we cannot directly move to cs. Instead we
+        // push the new segment selector and return value on the stack and use
+        // retfq to reload cs and continue at 1:.
+        #[cfg(feature = "inline_asm")]
+        asm!(
+            "push {sel}",
+            "lea {tmp}, [1f + rip]",
+            "push {tmp}",
+            "retfq",
+            "1:",
+            sel = in(reg) u64::from(sel.0),
+            tmp = lateout(reg) _,
+            options(preserves_flags),
+        );
+
+        #[cfg(not(feature = "inline_asm"))]
+        crate::asm::x86_64_asm_set_cs(u64::from(sel.0));
+    }
 }
 
-/// Reload gs segment register.
+/// Stack Segment
 ///
-/// ## Safety
+/// Entirely unused in 64-bit mode, setting the segment register does nothing.
+#[derive(Debug)]
+pub struct SS;
+segment_impl!(SS, "ss", x86_64_asm_get_ss, x86_64_asm_load_ss);
+
+/// Data Segment
 ///
-/// This function is unsafe because the caller must ensure that `sel`
-/// is a valid gs segment descriptor.
-#[inline]
-pub unsafe fn load_gs(sel: SegmentSelector) {
-    #[cfg(feature = "inline_asm")]
-    asm!("mov gs, {0:x}", in(reg) sel.0, options(nostack, preserves_flags));
+/// Entirely unused in 64-bit mode, setting the segment register does nothing.
```

Comment:
I think like the segment is still needed in user mode, so it's not entirely unused. There was some discussion about this in https://github.com/rust-osdev/x86_64/pull/78#issuecomment-508811071, but I don't remember all the details.

Reply:
This is a really good point that I missed; however, it doesn't deal with the DS segment, but the SS segment. Basically, the SS segment has to be set to a valid, non-null data segment in userspace, while the DS segment is still entirely unused (and can be null). For example, on my Linux system I have:
```rust
SS = SegmentSelector { index: 5, rpl: Ring3 }
CS = SegmentSelector { index: 6, rpl: Ring3 }
DS = SegmentSelector { index: 0, rpl: Ring0 }
ES = SegmentSelector { index: 0, rpl: Ring0 }
FS = SegmentSelector { index: 0, rpl: Ring0 }
GS = SegmentSelector { index: 0, rpl: Ring0 }
```

I need to read more about how/why SS is still used in long mode, and then update the `SS` docs.

Reply:
Thanks a lot for looking into this!

Reply:
I added a note to the SS doc section